client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java

```diff
@@ -455,25 +443,48 @@ public void flush() throws SegmentSealedException {
         }
     }
     
-    private void failConnection(Exception e) {
-        state.failConnection(e);
-        if (!state.isClosed() && state.reconnecting.compareAndSet(false, true)) {
+    private void failConnection(Throwable e) {
+        state.failConnection(ExceptionHelpers.getRealException(e));
+        reconnect();
+    }
+    
+    @VisibleForTesting
+    void reconnect() {
+        state.setupConnection.runReleaserAndAwait(() -> {
             Retry.indefinitelyWithExpBackoff(retrySchedule.getInitialMillis(), retrySchedule.getMultiplier(),
                                              retrySchedule.getMaxDelay(),
                                              t -> log.warn(writerId + " Failed to connect: ", t))
-                 .runInExecutor(() -> {
-                     if (!state.isClosed()) {
+                 .runAsync(() -> {
+                     if (state.isClosed() || state.isAlreadySealed()) {
+                         return CompletableFuture.completedFuture(null);
+                     }
+                     Preconditions.checkState(state.getConnection() == null);
+                     log.info("Fetching endpoint for segment {}", segmentName);
+                     return controller.getEndpointForSegment(segmentName).thenCompose((PravegaNodeUri uri) -> {
```

Comment:
You're executing this on the Executor on which `controller.getEndpointForSegment` was running. Is this acceptable? If this does significant sync work, then we might want to transfer it to our executor by calling `thenComposeAsync`